example/java/TestGenomicsDBImporterWithMergedVCFHeader.java

```diff
@@ -137,25 +144,50 @@ public static void main(final String[] args)
       files.add(args[i]);
     }
 
-    Map<String, FeatureReader<VariantContext>> map = new LinkedHashMap<>();
     List<VCFHeader> headers = new ArrayList<>();
+    ArrayList<String> sampleNames = new ArrayList<String>();
+    Map<String, String> sampleNameToFileName = new LinkedHashMap<String, String>();
 
+    //Get merged header first
     for (String file : files) {
       AbstractFeatureReader<VariantContext, LineIterator> reader =
         AbstractFeatureReader.getFeatureReader(file, new VCFCodec(), false);
-      String sampleName = ((VCFHeader) reader.getHeader()).getGenotypeSamples().get(0);
       headers.add((VCFHeader) reader.getHeader());
-      map.put(sampleName, reader);
+      final String sampleName = ((VCFHeader) reader.getHeader()).getGenotypeSamples().get(0);
+      sampleNames.add(sampleName);
+      sampleNameToFileName.put(sampleName, file);
+      //Hopefully, GC kicks in and frees resources assigned to reader
     }
-
     Set<VCFHeaderLine> mergedHeader = VCFUtils.smartMergeHeaders(headers, true);
 
-    GenomicsDBImporter importer = new GenomicsDBImporter(
-        map, mergedHeader,
-        new ChromosomeInterval(chromosomeName, Integer.parseInt(interval[0]), Integer.parseInt(interval[1])),
-        workspace, arrayName, 1000L, 1048576L,
-        useSamplesInOrder, failIfUpdating);
-    boolean isdone = importer.importBatch();
-    assert (isdone);
+    //This sorts the list sampleNames if !useSamplesInOrder
+    //Why you should use this? If you are writing multiple partitions in different machines,
+    //you must have consistent ordering of samples across partitions. If file order is different
+    //in different processes, then set useSamplesInOrder to false and let the sort in
+    //generateSortedCallSetMap ensure consistent ordering across samples
+    GenomicsDBImporter.generateSortedCallSetMap(sampleNames, useSamplesInOrder);
+
+    //Iterate over sorted sample list in batches
+    for(int i=0;i<sampleNames.size();i+=batchSize)
+    {
+      Map<String, FeatureReader<VariantContext>> map = new LinkedHashMap<>();
+      for(int j=i;j<sampleNames.size() && j<i+batchSize;++j)
+      {
+        final String sampleName = sampleNames.get(j);
+        assert sampleNameToFileName.containsKey(sampleName);
+        AbstractFeatureReader<VariantContext, LineIterator> reader =
+          AbstractFeatureReader.getFeatureReader(sampleNameToFileName.get(sampleName), new VCFCodec(), false);
+        assert sampleName.equals(((VCFHeader) reader.getHeader()).getGenotypeSamples().get(0));
+        map.put(sampleName, reader);
+      }
+      GenomicsDBImporter importer = new GenomicsDBImporter(
+          map, mergedHeader,
+          new ChromosomeInterval(chromosomeName, Integer.parseInt(interval[0]), Integer.parseInt(interval[1])),
+          workspace, arrayName, 1000L, 1048576L,
+          (long)i, (long)(i+batchSize-1),
```

Comment:
should be if (sampleNames.size()-i < batchSize) new importer with (i, i+(sampleNames.size()%batchSize)-1), else new importer with (i, i+batchSize-1)